### ISSUE TITLE
[Improvement]: Add list and remove operations to shuttle-persist

### ISSUE THREAD

**Issue by jonaro00** [shuttle-hq/shuttle#1052]
### Describe the improvement

The API of shuttle-persist consists of `save` and `load` methods.

Adding `list` (iterate over keys), and `remove` should be easy to add to the current implementation ~~so that I can finish my TODO list app FINALLY AFTER SIX MONTHS (jk)~~.

### Duplicate declaration

- [X] I have searched the issues and this improvement has not been requested before.

**Comment by sentinel1909** [comment]
I'll take this one, seems like a good starting point for me.  It will take me awhile though, and I might have to ask for some coaching along the way. :)

**Comment by sentinel1909** [comment]
Have been turning this issue over and around in my mind for the last couple of days. Trying to force myself to think about solutions before I start slinging code.  For the list method, I'd like to be able to iterate over all the keys to get their values. You can't iterate over the fields of a struct, so some Googling this morning revealed that the struct could be converted into something iterable, like an array.  Once you have an array, then you can iterate and do things.  This feels like a brute force approach, and I'm missing something more elegant, am I on the right track though?

**Comment by jonaro00** [comment]
Hey Jeff, here are some tips:

- (If the solution in my head works, which it might not 🤞) To solve this task, you need to add two public methods to the PersistInstace impl:
```rust
pub fn list(&self) -> Result<Vec<String>, PersistError> {
    ...
}

// "remove" or "delete", you choose the name
pub fn delete(&self) -> Result<(), PersistError> {
    ...
}

// bonus points: a remove_all/delete_all/clear that deletes everything
```
- Some things that will get you on the right track towards implementing them:
  - Where are keys & values saved? Where is the key string stored (and are there and bugs/limitations because of that?) and where is the value stored? (the data format is irrelevant for this task)
  - What type of operations are needed to achieve the tasks? (think which terminal commands they are similar to)
  - Which methods in the standard library can perform these operations? They return errors, so to handle that nicely, create two new variants of PersistError that represent and wrap that error type. Then, follow the same pattern as the save and load methods to convert the internal error to a PersistError.

Hope that helps! You can also ask in #build-together for more hints :D

**Comment by sentinel1909** [comment]
Thanks so much! Very helpful breadcrumbs for sure :)

**Comment by jonaro00** [comment]
The save function takes in `key: &str` and `struc: T`, where the generic `<T: Serialize>` specifies that `T` must implement `Serialize`. Why? So that it can be saved and later deserialized. (load has the same constraints but reversed)

`T` can be any type, not only structs (for example, native types `String`, `i32`, etc, but also Enums)
Most native types already are `Serialize`, and as you know, you can `#[derive(Serialize)]` to make your own structs/enums implement it.

So in summary, the persist storage can be described as:
A persistent key-value store, where the key is `&str` and the value is any type `T` that implements `Serialize` and `DeserializeOwned`.

**Comment by sentinel1909** [comment]
I'm focusing on the list functionality first.  I'm imagining that each PersistInstance can have a folder associated with it and then within each folder is a file, where the file name is in the format {key}.bin.  Within the file is the data.  So, in wanting to list, I'd need to do something like an ls terminal command on the folder.  I'm envisioning iterating over the contents of the folder, to get the file path as a string, and storing it all in a Vector to return for the user to do with what they will.  TBH I'm having a little trouble, just by reading and thinking through what the code is doing, with the relationsip between a PersistInstance and a folder.

**Comment by jonaro00** [comment]
You are correct that it is rooted in some folder (where this is placed is unrelated to this task, and the name "instance" is a bit weird since different instances will operate on the same folder) and that the listing task can be solved by iterating over the directory entries in the directory returned by the private method.

**Comment by sentinel1909** [comment]
One last question (hopefully)... I'm uncertain as to what the get_storage_folder function is doing.  I understand that it returns a PathBuf,  but am confused by what the code `["shuttle_persist", &self.service_name.to_string()].iter().collect()` is doing.
